### PR TITLE
create mongodb aggregation framework article

### PR DESCRIPTION
**Pull request by alex-emerich**

### CODE REVIEW

File: content/08-mongodb/18-mongodb-aggregation-framework.mdx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	{ _id: 5, first_name: "Henry David", last_name: "Thoreau", book_title: 'Walden', genre: 'Nonfiction', quantity: 15, date: ISODate( "2021-03-13T11:19:30Z" ) },
+	{ _id: 6, first_name: "Virginia", last_name: "Woolf", book_title: "A Room of One's Own", genre: 'Nonfiction',
+     quantity: 18, date: ISODate( "2022-10-21T11:19:30Z" ) },
+	{ _id: 7, first_name: "Virginia", last_name: "Woolf", book_title: "Mr's Dalloway", genre: 'Fiction', quantity: 14, date: ISODate( "2022-10-21T11:19:30Z" ) },
+	{ _id: 8, first_name: "Zadie", last_name: "Smith", book_title: 'White Teeth', genre: 'Fiction', quantity: 8, date: ISODate( "2022-10-21T11:19:30Z" ) },
+	{ _id: 9, first_name: "Charles", last_name: "Dickens", book_title: 'The Old Curiousity Shop', genre: 'Fiction', quantity: 6, date: ISODate( "2022-10-21T11:19:30Z" ) }
+] )
+```
+
+Now that our collection has some sample documents, we can start our query. Aggregation pipelines run with the `db.<collection-name>.aggregate()` method.  Our goal is to design a query that return a list of the authors with the most total copies of their fiction books ordered. An example aggregation query can be found below with each stage described.
```
Comment (imchairmanm) [inline content/08-mongodb/18-mongodb-aggregation-framework.mdx:64]:
```suggestion
Now that our collection has some sample documents, we can start our query. Aggregation pipelines run with the `db.<collection-name>.aggregate()` method.  Our goal is to design a query that returns a list of the authors with the most total copies of their fiction books ordered. An example aggregation query can be found below with each stage described. 
```

File: content/08-mongodb/18-mongodb-aggregation-framework.mdx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	{ _id: 8, first_name: "Zadie", last_name: "Smith", book_title: 'White Teeth', genre: 'Fiction', quantity: 8, date: ISODate( "2022-10-21T11:19:30Z" ) },
+	{ _id: 9, first_name: "Charles", last_name: "Dickens", book_title: 'The Old Curiousity Shop', genre: 'Fiction', quantity: 6, date: ISODate( "2022-10-21T11:19:30Z" ) }
+] )
+```
+
+Now that our collection has some sample documents, we can start our query. Aggregation pipelines run with the `db.<collection-name>.aggregate()` method.  Our goal is to design a query that return a list of the authors with the most total copies of their fiction books ordered. An example aggregation query can be found below with each stage described. 
+
+```
+db.bookOrders.aggregate ( [ 
+	// Stage 1: The $match operator scans the collection for documents
```
Comment (imchairmanm) [inline content/08-mongodb/18-mongodb-aggregation-framework.mdx:68]: Really like the use of comments here

File: content/08-mongodb/18-mongodb-aggregation-framework.mdx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  { _id: 'Smith', totalQuantity: 8 },
+  { _id: 'Derrida', totalQuantity: 5 }
+]
+```
+
+This example is intentionally simple, but it demonstrates how an aggregation pipeline can take some of the complexity out of some queries. Each step to reaching your desired output is clearly broken down and compartmentalized into a clear stage.
+
+Depending on the collection and document data structure, there are optmizations to consider when building an aggregation pipeline. Additionally, this framework may not work for all complex logic. It is case dependent. 
+
+One small optimization that should be pointed out can be seen in the first two stages of our example. Generally, the `$match` operator is used to begin most pipelines and best practice. However, if your collection is full of very large documents, then it is recommended to begin with the `$project` operator instead. Starting with `$project` limits the amount of fields that get passed onto the next stage earlier in the pipeline and reduces some unnecessary load.
```
Comment (imchairmanm) [inline content/08-mongodb/18-mongodb-aggregation-framework.mdx:122]: This is a really great tip

**Review by imchairmanm — APPROVED** [review]
Really, really cool article.  A+

One thing I thought about as I was reading was how this type of framework encourages composition and iteration.  You can continually add stages to your pipeline to finesse the data until it's in the appropriate shape.  Coming from *nix and working with shells, I think that's a really powerful concept that's really easy to reason about.

File: content/08-mongodb/18-mongodb-aggregation-framework.mdx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+With an aggregation framework, one can achieve several goals. We’ll go into specific examples with the actual operation syntax, but in theory an analyst for the fiction department at a bookstore could set up a framework that groups the number of purchases based on genre or author to inform the sales floor. No matter the team, there are insights to be had from data that are all the more easily discovered with the breakdown done by an aggregation pipeline.
+
+## What are the most common MongoDB aggregation operations?
+
+There are approximately 38 aggregation stages available in the MongoDB framework at the time of this writing. We are not going to delve into all of them in this guide, but you can view the whole list in the [official MongoDB documentation](https://www.mongodb.com/docs/manual/reference/operator/aggregation-pipeline/#alphabetical-listing-of-stages). We’ll spend some time to highlight a few that will also get used in an example pipeline.
```
[the remark this answers is not in the document]
Comment (alex-emerich) [reply content/08-mongodb/18-mongodb-aggregation-framework.mdx:32]: XD

File: content/08-mongodb/18-mongodb-aggregation-framework.mdx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+- `$project` : Reshapes each document in the stream, such as by adding new fields or removing existing fields. For each input document, outputs one document.
+- `$match` : Filters the document stream to allow only matching documents to pass unmodified into the next pipeline stage. `$match` uses standard MongoDB queries. For each input document, outputs either one document (a match) or zero documents (no match). 
+- `$group` : Groups input documents by a specified identifier expression and applies the accumulator expression(s), if specified, to each group. Consumes all input documents and outputs one document per each distinct group. The output documents only contain the identifier field and, if specified, accumulated fields.
+- `$sort` : Reorders the document stream by a specified sort key. Only the order changes; the documents remain unmodified. For each input document, outputs one document. 
+- `$skip` : Skips the first n documents where n is the specified skip number and passes the remaining documents unmodified to the pipeline. For each input document, outputs either zero documents (for the first n documents) or one document (if after the first n documents).
```
[the remark this answers is not in the document]
Comment (alex-emerich) [reply content/08-mongodb/18-mongodb-aggregation-framework.mdx:38]: Ah yes. Good call

**Pull request merged**